portforward/forwarder.go

```diff
@@ -0,0 +1,272 @@
+	f.cancel = cancel
+
+	logger := f.opts.Logger.With(
+		slog.F("network", f.spec.Network),
+		slog.F("listen_host", f.spec.ListenHost),
+		slog.F("listen_port", f.spec.ListenPort),
+	)
+
+	listenAddress := netip.AddrPortFrom(f.spec.ListenHost, f.spec.ListenPort)
+	dialAddress := fmt.Sprintf("127.0.0.1:%d", f.spec.DialPort)
+
+	l, err := f.opts.Listener.Listen(f.spec.Network, listenAddress.String())
+	if err != nil {
+		cancel()
+		return xerrors.Errorf("listen '%s://%s': %w", f.spec.Network, listenAddress.String(), err)
+	}
+	f.listener = l
+	logger.Debug(ctx, "listening")
+
+	f.active.Store(true)
+
+	f.wg.Add(1)
+	go func() {
+		defer func() {
+			f.wg.Done()
+			f.active.Store(false)
+		}()
+
+		for {
+			netConn, err := l.Accept()
+			if err != nil {
+				// Silently ignore net.ErrClosed errors.
+				if xerrors.Is(err, net.ErrClosed) {
+					logger.Debug(ctx, "listener closed")
+					return
+				}
+				logger.Error(ctx, "error accepting connection",
+					slog.F("listen_address", listenAddress.String()),
+					slog.Error(err))
+				return
+			}
+			logger.Debug(ctx, "accepted connection",
+				slog.F("remote_addr", netConn.RemoteAddr()))
+
+			go func(netConn net.Conn) {
+				defer netConn.Close()
+				remoteConn, err := f.opts.Dialer.DialContext(ctx, f.spec.Network, dialAddress)
+				if err != nil {
+					logger.Error(ctx, "failed to dial remote",
+						slog.F("dial_address", dialAddress),
+						slog.Error(err))
+					return
+				}
+				defer remoteConn.Close()
+				logger.Debug(ctx, "dialed remote",
+					slog.F("remote_addr", netConn.RemoteAddr()))
+
+				agentssh.Bicopy(ctx, netConn, remoteConn)
+				logger.Debug(ctx, "connection closing",
+					slog.F("remote_addr", netConn.RemoteAddr()))
+			}(netConn)
+		}
+	}()
+
+	return nil
+}
+
+func (f *LocalForwarder) Stop() error {
+	if !f.active.Load() {
+		return nil
+	}
+
+	if f.cancel != nil {
+		f.cancel()
+	}
+	if f.listener != nil {
+		_ = f.listener.Close()
+	}
+	f.wg.Wait()
+	return nil
+}
+
+func (f *LocalForwarder) IsActive() bool {
+	return f.active.Load()
+}
+
+func (f *LocalForwarder) Spec() Spec {
+	return f.spec
+}
+
+// manager implements the Manager interface.
+type manager struct {
+	forwarders map[string]Forwarder
+	opts       Options
+	mu         sync.RWMutex
+}
+
+// NewManager creates a new port forwarding manager.
+func NewManager(opts Options) Manager {
+	return &manager{
+		forwarders: make(map[string]Forwarder),
+		opts:       opts,
+	}
+}
+
+func (m *manager) Add(spec Spec) (Forwarder, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	key := fmt.Sprintf("%s:%s:%d", spec.Network, spec.ListenHost, spec.ListenPort)
+	if _, exists := m.forwarders[key]; exists {
+		return nil, xerrors.Errorf("forwarder already exists for %s", key)
+	}
+
+	// Test if we can actually bind to the port before adding the forwarder
+	listenAddress := netip.AddrPortFrom(spec.ListenHost, spec.ListenPort)
+	testListener, err := m.opts.Listener.Listen(spec.Network, listenAddress.String())
```

Comment:
`Add()` listens but doesn't start forwarding? IMO the `Start()` API should just be removed in favor of `Add()` always creating the listener and starting the forwarding goroutine